nexus/src/authn/silos.rs

```diff
@@ -0,0 +1,203 @@
+use samael::metadata::EntityDescriptor;
+use samael::metadata::NameIdFormat;
+use samael::metadata::HTTP_REDIRECT_BINDING;
+use samael::service_provider::ServiceProvider;
+use samael::service_provider::ServiceProviderBuilder;
+
+pub struct SamlIdentityProvider {
+    pub idp_metadata_document_string: String,
+    pub sp_client_id: String,
+    pub acs_url: String,
+    pub slo_url: String,
+    pub technical_contact_email: String,
+    pub public_cert: Option<String>,
+    pub private_key: Option<String>,
+}
+
+impl TryFrom<model::SamlIdentityProvider> for SamlIdentityProvider {
+    type Error = anyhow::Error;
+    fn try_from(
+        model: model::SamlIdentityProvider,
+    ) -> Result<Self, Self::Error> {
+        let provider = SamlIdentityProvider {
+            idp_metadata_document_string: model.idp_metadata_document_string,
+            sp_client_id: model.sp_client_id,
+            acs_url: model.acs_url,
+            slo_url: model.slo_url,
+            technical_contact_email: model.technical_contact_email,
+            public_cert: model.public_cert,
+            private_key: model.private_key,
+        };
+
+        // check that the idp metadata document string parses into an EntityDescriptor
+        let _idp_metadata: EntityDescriptor =
+            provider.idp_metadata_document_string.parse()?;
+
+        // check that there is a valid sign in url
+        let _sign_in_url = provider.sign_in_url(None)?;
+
+        Ok(provider)
+    }
+}
+
+pub enum IdentityProviderType {
+    Saml(SamlIdentityProvider),
+}
+
+impl IdentityProviderType {
+    /// First, look up the provider type, then look in for the specific
+    /// provider details.
+    pub async fn lookup(
+        datastore: &DataStore,
+        opctx: &OpContext,
+        silo_name: &model::Name,
+        provider_name: &model::Name,
+    ) -> LookupResult<Self> {
+        let (.., identity_provider) = LookupPath::new(opctx, datastore)
+            .silo_name(silo_name)
+            .identity_provider_name(provider_name)
+            .fetch()
+            .await?;
+
+        match identity_provider.provider_type {
+            model::IdentityProviderType::Saml => {
+                let (.., saml_identity_provider) =
+                    LookupPath::new(opctx, datastore)
+                        .silo_name(silo_name)
+                        .saml_identity_provider_name(provider_name)
+                        .fetch()
+                        .await?;
+
+                Ok(IdentityProviderType::Saml(
+                    saml_identity_provider.try_into()
+                        .map_err(|e: anyhow::Error|
+                            // If an error is encountered converting from the
+                            // model to the authn type here, this is a server
+                            // error: it was validated before it went into the
+                            // DB.
+                            omicron_common::api::external::Error::internal_error(&e.to_string())
```

Comment:
Can you add a tiny bit of context here (like maybe make that argument `&format!("deserializing SAML IdP from database: {:#}", e)`?  I'm afraid if we ever hit this case it'll be very hard to tell what we were even trying to parse.

Reply:
Definitely, context added in 46de770d